upcoming: [M3-9111] - Add Upgrade Interfaces dialog

coliu-akamai opened the pull request:
## Description 📝
- Adds the Upgrade Interface dialog. Linodes with Legacy Config Profile interfaces can have their interfaces upgraded to new Linode Interfaces

## Changes  🔄
- Add components and unit tests for dialog - interactive tests will be built out in M3-9176
- add relevant query
- updating some props for the button component to always show a tooltip
- Added updateInterface handler in MSW (haven't handled firewall stuff in this msw)

## Preview 📷
This video is from testing on prod, so there's a not found error (having trouble with devenv) - it does give an example of the error flow though!
(there may be a few more UI tweaks to the dialog after this video)

https://github.com/user-attachments/assets/a73cd320-0697-4b1b-87c5-7c11738bd2d2

## How to test 🧪

### Prerequisites
- on devenv OR with MSW CRUD:
- have a Linode using legacy configs with one config
- have a Linode using legacy configs with multiple configs
- have a Linode using new Linode Interfaces

On prod: you won't be able to test the endpoint working, but you can test errors
On MSW: you can test success states

Using the MSW: you can seed a Linode (will have configuration profile interfaces) and test that upgrading "works". To test multiple configs, you can update the `createLinode` handler at line 224 to include another config, then create a linode:
```
    const linodeConfig2 = configFactory.build({
      created: DateTime.now().toISO(),
    });

    await mswDB.add('linodes', linode, mockState);
    await mswDB.add('linodeConfigs', [linode.id, linodeConfig], mockState);
    await mswDB.add('linodeConfigs', [linode.id, linodeConfig2], mockState);
```
![image](https://github.com/user-attachments/assets/b76ea7cb-eb7f-4891-b280-b34ca1ff3aa2)


### Verification steps
- Confirm everything is hidden if feature flag is off
- Confirm the Upgrade Button / chip don't show up for a Linode with new interfaces
- Confirm flow for a Linode with only one config (multiple config select shouldn't show up)
  - Confirm both dry run and actual upgrade
  - Confirm error flow for dry run and actual upgrade (ex: Account settings for interfaces is "Config Profile only") (must test with devenv)
- Confirm flow for a Linode with multiple configs
  - Confirm both dry run and actual upgrade
  - Confirm error flow for dry run and actual upgrade (ex: Account settings for interfaces is "Config Profile only") (must test with devenv)

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

bnussman-akamai commented on packages/manager/src/features/Linodes/LinodeEntityDetailBody.tsx:
```diff
@@ -446,11 +452,32 @@ export const LinodeEntityDetailBody = React.memo((props: BodyProps) => {
           )}
         </Grid>
       )}
+      <UpgradeInterfacesDialog
+        linodeId={linodeId}
+        onClose={() => setIsUpgradeInterfacesDialogOpen(false)}
+        open={isUpgradeInterfacesDialogOpen}
+      />
```
Hmm. I think this is okay as is, but we want to try, maybe we could try placing the dialog at a higher level and use a deep-link to manage the open/close state to prevent prop drilling?

bnussman-akamai commented on packages/manager/src/features/Linodes/LinodesDetail/LinodeConfigs/UpgradeInterfaces/DialogContents/ConfigSelectDialogContent.tsx:
```diff
@@ -0,0 +1,75 @@
+        options={configOptions}
+        placeholder="Select Configuration Profile"
+      />
+      <Stack direction="row-reverse" gap={2} sx={{ marginTop: 4 }}>
+        <Button
+          buttonType="primary"
+          disabled={!selectedConfigId}
+          onClick={() => upgradeToLinodeInterfaces(state.isDryRun)}
+        >
+          {state.isDryRun ? 'Upgrade Dry Run' : 'Upgrade Interfaces'}
+        </Button>
+        <Button buttonType="secondary" onClick={onClose}>
+          Cancel
+        </Button>
+      </Stack>
```
Could we switch the order of the buttons and use `row` instead of `row-reverse`. I guess it's the same thing, it just feels a bit backwards to me

```suggestion
      <Stack direction="row" gap={2} sx={{ marginTop: 4 }}>
        <Button buttonType="secondary" onClick={onClose}>
          Cancel
        </Button>
        <Button
          buttonType="primary"
          disabled={!selectedConfigId}
          onClick={() => upgradeToLinodeInterfaces(state.isDryRun)}
        >
          {state.isDryRun ? 'Upgrade Dry Run' : 'Upgrade Interfaces'}
        </Button>
      </Stack>
```

coliu-akamai replied:
will do! switching to box/flex - row started moved all the buttons to the left side

coliu-akamai commented on packages/manager/src/features/Linodes/LinodesDetail/LinodeConfigs/UpgradeInterfaces/DialogContents/ErrorDialogContent.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <Stack gap={2}>
+      <Notice important variant="error">
+        <Typography>
+          {isDryRun
+            ? ERROR_DRY_RUN_COPY
+            : errors.length > 0
+            ? errors[0].reason
+            : 'An unexpected error occurred.'}
+        </Typography>
```
any preference?

| current  | proposed - matches dry run error appearance   |
| ------- | ------- |
| ![image](https://github.com/user-attachments/assets/225941b3-a177-4efb-9ef3-b8231f134241) | ![image](https://github.com/user-attachments/assets/ff7b8ca6-1ab3-408f-9ba5-7d9b71779b35) |

(also I just pushed up a small change to add some padding so that the button isn't so close to the top of the dialog in this state...I may get rid of that tbh)

bnussman-akamai replied:
I like the **proposed** because we _could_ have more than 1 error, right?

bnussman-akamai commented on packages/manager/src/features/Linodes/LinodesDetail/LinodeConfigs/UpgradeInterfaces/useUpgradeToLinodeInterfaces.ts:
```diff
@@ -0,0 +1,63 @@
+          step: 'progress',
+        });
+        // return control to the DOM to update the progress
+        await new Promise((resolve) => setTimeout(resolve, 0));
+        setDialogState({
+          dialogTitle,
+          isDryRun,
+          progress: 80,
+          step: 'progress',
+        });
```
I know the mockups use this loading state pattern, but I'd be in favor of just using button loading states and dropping the LinearProgress pattern. In my opinion, it's a bit extra, but up to you and/or UX.

bnussman-akamai reviewed (commented):
Looking solid so far, going to do a bit more testing

coliu-akamai commented:
@bnussman-akamai should be ready for a re-review!
- Added deep-linking, lmk your thoughts on it! It feels a bit brittle?? but I think this works given all the different routes we have for Linodes
- removed the progress state after UX review

bnussman-akamai commented on packages/manager/src/features/Linodes/LinodeEntityDetailBody.tsx:
I can't tell for sure, but it looks like the chip and the text might not be vertically centered

![Screenshot 2025-03-17 at 11 19 36 AM](https://github.com/user-attachments/assets/33618737-0f9d-4ba8-ad04-16ea3d492585)

carrillo-erik reviewed (approved):
I was able to verify the changes using the MSW and Feature Flags in the CM dev tools. The text within the `Upgrade` chip does seem to be vertically misaligned as noted by @bnussman-akamai.
